### PR TITLE
ABC-91 Add server request to emoji autocomplete

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
Add server request to emoji autocomplete. The diff looks like I did a lot but really I just encapsulated some functionality into a function and added the server request.

@jasonblais this is going to suffer from the same "server results popping in" issue that the other autocomplete functions have

#### Ticket Link
https://mattermost.atlassian.net/browse/ABC-91

### CODE REVIEW

File: components/suggestion/emoticon_provider.jsx [hunk 1]
```diff
@@ -49,82 +53,88 @@ class EmoticonSuggestion extends Suggestion {
...
-            SuggestionStore.clearSuggestions(suggestionId);
-            if (terms.length > 0) {
-                SuggestionStore.addSuggestions(suggestionId, terms, matched, EmoticonSuggestion, text);
+        const emojiMap = getEmojiMap(store.getState());
 
-                hasSuggestions = true;
+        // Check for named emoji
+        for (const [name, emoji] of emojiMap) {
+            if (emoji.aliases) {
+                // This is a system emoji so it may have multiple names
```
Comment (stephenkiers) [inline components/suggestion/emoticon_provider.jsx]: Why is comment inside if block here...

File: components/suggestion/emoticon_provider.jsx [hunk 2]
```diff
@@ -49,82 +53,88 @@ class EmoticonSuggestion extends Suggestion {
...
+                // This is a custom emoji so it only has one name
+                matched.push({name, emoji});
             }
         }
 
-        if (hasSuggestions) {
-            // force the selection to be cleared since the order of elements may have changed
-            SuggestionStore.clearSelection(suggestionId);
+        // Sort the emoticons so that emoticons starting with the entered text come first
+        matched.sort((a, b) => {
```
Comment (stephenkiers) [inline components/suggestion/emoticon_provider.jsx]: I feel like this method should be moved out of here, for easier reading.

    matched = sortWithPreferenceGivenToPrefixMatches(matched, partialName);

**Comment by jasonblais** [comment]
@jwilander Cannot type more than one character following a comma, and hence can't test the autocomplete. 
image

**Comment by jwilander** [comment]
Oh sorry, this isn't testable yet. It needs some redux PRs to go in first

**Comment by jwilander** [comment]
@jasonblais This one should be ready for testing and has a bunch of custom emoji. For some reason the spinmints weren't getting the correct redux commits so I had to manually mess with them to get it to work

**Comment by jwilander** [comment]
I should also note that I disabled loading all emoji on page load so that this can be properly tested, That does mean that posts might show the text instead of the custom emoji, but that's expected and won't happen when it's merged.

**Comment by jasonblais** [comment]
If I throttle my connection to a lower network speed, there's definitely a degradation in the user experience. The worry I have is that it's worse here, given the emoji I was looking at may get hidden after all the custom emoji are loaded (whereas for channels, the client-side results are always listed first, and the focus isn't lost).

Let's discuss during our 1-1

**Comment by jasonblais** [comment]
I'll post about this in UX channel to get other people's opinions.

**Comment by jasonblais** [comment]
Should hear back soon, I think we either go with the current implementation, or always load local results first before server-side results.

**Comment by jasonblais** [comment]
@jwilander UX team leaning towards the implementation we have now. Question though:

>  Is there no way of caching the data beyond a single session? So that it doesn't have to pull down a fresh copy every time i refresh if its not necessary? It would be good to reduce how frequently someone would end up encountering the issue.

**Comment by jwilander** [comment]
Yes, it's possible to do that. The RN apps do that. It'll be more work though

**Pull request merged**